File: WellnessWingman/Platforms/Android/Services/AnalysisForegroundService.cs
using Android.App;
using Android.Content;
using Android.Content.PM;
using Android.OS;
using AndroidX.Core.App;

namespace HealthHelper.Platforms.Android.Services;

/// <summary>
/// Android foreground service that ensures LLM analysis completes even when
/// the screen is locked or the app is backgrounded.
/// Shows a persistent notification while analysis is running.
/// </summary>
[Service(Exported = false, ForegroundServiceType = ForegroundService.TypeDataSync)]
public class AnalysisForegroundService : Service
{
    private const string ChannelId = "analysis_channel";
    private const string ChannelName = "Analysis";
    private const int NotificationId = 1001;
    private static readonly object _lock = new object();
    private static int _activeTaskCount = 0;

    public override IBinder? OnBind(Intent? intent)
    {
        // Not a bound service
        return null;
    }

    public override StartCommandResult OnStartCommand(Intent? intent, StartCommandFlags flags, int startId)
    {
        lock (_lock)
        {
            _activeTaskCount++;

            // Create notification channel (required for Android 8+)
            CreateNotificationChannel();

            // Build and display notification
            var notification = BuildNotification();
            StartForeground(NotificationId, notification);
        }

        // Return Sticky so service is restarted if killed by system
        return StartCommandResult.Sticky;
    }

    public override void OnDestroy()
    {
        lock (_lock)
        {
            _activeTaskCount--;

            if (_activeTaskCount <= 0)
            {
                _activeTaskCount = 0;
                if (OperatingSystem.IsAndroidVersionAtLeast(24))
                {
                    StopForeground(StopForegroundFlags.Remove);
                }
                else
                {
#pragma warning disable CS0618 // StopForeground(bool) is obsolete but required for API < 24
                    StopForeground(true);
#pragma warning restore CS0618
                }
            }
        }

        base.OnDestroy();
    }

    /// <summary>
    /// Decrements the task count and stops the service if no tasks remain.
    /// Called by AndroidBackgroundExecutionService when a task completes.
    /// </summary>
    public static void DecrementTaskCount(Context context)
    {
        lock (_lock)
        {
            _activeTaskCount--;

            if (_activeTaskCount <= 0)
            {
                _activeTaskCount = 0;
                var intent = new Intent(context, typeof(AnalysisForegroundService));
                context.StopService(intent);
            }
        }
    }

    private void CreateNotificationChannel()
    {
        if (OperatingSystem.IsAndroidVersionAtLeast(26))
        {
            var channel = new NotificationChannel(
                ChannelId,
                ChannelName,
                NotificationImportance.Low) // Low = no sound/vibration
            {
                Description = "Shows when analyzing meal photos"
            };

            var notificationManager = GetSystemService(NotificationService) as NotificationManager;
            notificationManager?.CreateNotificationChannel(channel);
        }
    }

    private Notification BuildNotification()
    {
        // Create intent to open app when notification is tapped
        var pendingIntentFlags = OperatingSystem.IsAndroidVersionAtLeast(23)
            ? PendingIntentFlags.Immutable
            : PendingIntentFlags.UpdateCurrent;

        var pendingIntent = PendingIntent.GetActivity(
            this,
            0,
            PackageManager?.GetLaunchIntentForPackage(PackageName ?? string.Empty),
            pendingIntentFlags);

        var taskCountText = _activeTaskCount == 1
            ? "Processing 1 entry..."
            : $"Processing {_activeTaskCount} entries...";

        var builder = new NotificationCompat.Builder(this, ChannelId)!
            .SetContentTitle("Analyzing your entries")!
            .SetSmallIcon(global::Android.Resource.Drawable.IcMenuUpload)! // Using system icon for now
            .SetOngoing(true)! // Makes it persistent
            .SetPriority(NotificationCompat.PriorityLow)!; // Low priority = no sound

        if (pendingIntent != null)
        {
            builder.SetContentIntent(pendingIntent);
        }

        if (taskCountText != null)
        {
            builder.SetContentText(taskCountText);
        }

        return builder.Build()!;
    }
}

File: kotlin/shared/src/androidMain/kotlin/com/wellnesswingman/service/AnalysisForegroundService.kt
package com.wellnesswingman.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import io.github.aakira.napier.Napier

/**
 * Android foreground service that ensures LLM analysis completes even when
 * the screen is locked or the app is backgrounded.
 * Shows a persistent notification while analysis is running.
 */
class AnalysisForegroundService : Service() {

    override fun onBind(intent: Intent?): IBinder? {
        // Not a bound service
        return null
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        synchronized(lock) {
            activeTaskCount++
            Napier.d("Foreground service started. Active tasks: $activeTaskCount")

            // Create notification channel (required for Android 8+)
            createNotificationChannel()

            // Build and display notification
            val notification = buildNotification()
            startForeground(NOTIFICATION_ID, notification)
        }

        // Return Sticky so service is restarted if killed by system
        return START_STICKY
    }

    override fun onDestroy() {
        synchronized(lock) {
            activeTaskCount--
            Napier.d("Foreground service destroyed. Active tasks: $activeTaskCount")

            if (activeTaskCount <= 0) {
                activeTaskCount = 0
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    stopForeground(STOP_FOREGROUND_REMOVE)
                } else {
                    @Suppress("DEPRECATION")
                    stopForeground(true)
                }
            }
        }

        super.onDestroy()
    }

    private fun createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                CHANNEL_NAME,
                NotificationManager.IMPORTANCE_LOW // Low = no sound/vibration
            ).apply {
                description = "Shows when analyzing meal photos"
            }

            val notificationManager = getSystemService(NotificationManager::class.java)
            notificationManager?.createNotificationChannel(channel)
        }
    }

    private fun buildNotification(): Notification {
        // Create intent to open app when notification is tapped
        val pendingIntentFlags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PendingIntent.FLAG_IMMUTABLE
        } else {
            PendingIntent.FLAG_UPDATE_CURRENT
        }

        val launchIntent = packageManager.getLaunchIntentForPackage(packageName)
        val pendingIntent = launchIntent?.let {
            PendingIntent.getActivity(this, 0, it, pendingIntentFlags)
        }

        val taskCountText = if (activeTaskCount == 1) {
            "Processing 1 entry..."
        } else {
            "Processing $activeTaskCount entries..."
        }

        val builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            Notification.Builder(this, CHANNEL_ID)
        } else {
            @Suppress("DEPRECATION")
            Notification.Builder(this)
                .setPriority(Notification.PRIORITY_LOW)
        }

        builder
            .setContentTitle("Analyzing your entries")
            .setContentText(taskCountText)
            .setSmallIcon(android.R.drawable.ic_menu_upload) // Using system icon for now
            .setOngoing(true) // Makes it persistent

        pendingIntent?.let { builder.setContentIntent(it) }

        return builder.build()
    }

    companion object {
        const val CHANNEL_ID = "analysis_channel"
        const val CHANNEL_NAME = "Analysis"
        const val NOTIFICATION_ID = 1001
        const val EXTRA_TASK_NAME = "taskName"

        private val lock = Any()
        private var activeTaskCount = 0

        /**
         * Decrements the task count and stops the service if no tasks remain.
         * Called by AndroidBackgroundExecutionService when a task completes.
         */
        fun decrementTaskCount(context: Context) {
            synchronized(lock) {
                activeTaskCount--
                Napier.d("Task decremented. Active tasks: $activeTaskCount")

                if (activeTaskCount <= 0) {
                    activeTaskCount = 0
                    val intent = Intent(context, AnalysisForegroundService::class.java)
                    context.stopService(intent)
                    Napier.d("Stopping foreground service - no active tasks")
                }
            }
        }
    }
}
